mbbsd/mail.c

```diff
@@ -1213,42 +1213,96 @@ get_account_sysop(struct Vector *namelist)
         return -1;
 }
 
+void
+show_mail_account_sysop_desc(void)
+{
+    more(FN_MAIL_ACCOUNT_SYSOP_DESC, 1);
+}
+
+const char *
+select_account_sysop(struct Vector *namelist)
+{
+    int i, sysop_index;
+    char linebuf[STRLEN+1];
+
+    vs_hdr("�п�ܭn�H�����@��b������");
+    for (i = 0; i < Vector_length(namelist) && i + 2 < t_lines; i++) {
+        const char *userid = Vector_get(namelist, i);
+        snprintf(linebuf, sizeof(linebuf), "%2d. %s", i + 1, userid);
+        move(i + 2, 0);
+        outs(linebuf);
+    }
+    getdata(1, 0, "�п�J�b������s��: ", linebuf, 4, LCECHO);
+
+    // This implementation allows users to select the account sysop
+    // who is not showed on the screen due to the limitation of screen size
+    if (sscanf(linebuf, "%d", &sysop_index) != 1 || sysop_index < 1 || sysop_index > Vector_length(namelist))
+        return NULL;
+
+    return Vector_get(namelist, sysop_index - 1);
+}
 
 int
 mail_account_sysop(void)
 {
     // send mail to account sysop
-    int             i;
     fileheader_t    mymail;
     char            fpath[TTLEN];
     char            genbuf[PATHLEN];
     char            idbuf[IDLEN+1];
     int             oldstat = currstat;
     char            save_title[STRLEN];
-    char            tmp_title[STRLEN-20];
+    char            tmp_title[STRLEN];
     struct          Vector namelist;
+    char            linebuf[STRLEN+1];
+    const char     *userid;
 
     Vector_init(&namelist, IDLEN+1);
 
     if (get_account_sysop(&namelist) == -1)
     {
-        vmsg("Ū���H�W�楢�ѡA�ЦV " BN_BUGREPORT " �^��");
+        vmsg("Ū���H�W�楢�ѡA�ЦV " BN_SYSOP " �^��");
         Vector_delete(&namelist);
         return DIRCHANGED;
     }
 
     if (Vector_length(&namelist) == 0)
     {
-        vmsg("�L���Ħ���H�W��A�ЦV " BN_BUGREPORT " �^��");
+        vmsg("�L���Ħ���H�W��A�ЦV " BN_SYSOP " �^��");
         Vector_delete(&namelist);
         return DIRCHANGED;
     }
 
-    getdata(2, 0, "�D�D:", tmp_title, sizeof(tmp_title), DOECHO);
-    strlcpy(save_title, tmp_title, sizeof(save_title));
+    clear();
+    show_mail_account_sysop_desc();
 
     setutmpmode(SMAIL);
```

Comment:
I think that is probably not a good idea to put the error handling of "get_account_sysop" and "select_account_sysop" together, because the former is about the error of getting the ID list of account sysops that should be reported, but the latter is about the error of selecting an invalid sequence number of account sysops that should be corrected by the user himself. So, the error messages of these two situations should be different.